service/src/main/java/uk/nhs/adaptors/gp2gp/common/task/TaskErrorHandler.java

```diff
@@ -18,26 +19,33 @@
 import uk.nhs.adaptors.gp2gp.gpc.exception.GpConnectException;
 import uk.nhs.adaptors.gp2gp.gpc.exception.GpConnectInvalidException;
 import uk.nhs.adaptors.gp2gp.gpc.exception.GpConnectNotFoundException;
+import uk.nhs.adaptors.gp2gp.gpc.exception.GpcServerErrorException;
 
 @Component
 @Slf4j
 @AllArgsConstructor(onConstructor = @__(@Autowired))
 public class TaskErrorHandler {
 
-    private final Map<Class<? extends Exception>, Function<TaskDefinition, Boolean>> errorHandlers = Map.of(
+    private final Map<Class<? extends Throwable>, Function<TaskDefinition, Boolean>> errorHandlers = Map.of(
         EhrRequestException.class, this::handleRequestError,
         EhrExtractException.class, this::handleTranslationError,
         EhrMapperException.class, this::handleTranslationError,
         FhirValidationException.class, this::handleTranslationError,
         GpConnectException.class, this::handleGpConnectError,
         GpConnectInvalidException.class, this::handleInvalidNotAuthError,
         GpConnectNotFoundException.class, this::handleNotFoundError,
-        MaximumExternalAttachmentsException.class, this::handleMaximumExternalAttachmentsError
+        MaximumExternalAttachmentsException.class, this::handleMaximumExternalAttachmentsError,
+        GpcServerErrorException.class, this::handleGpConnectError
     );
 
     private final ProcessFailureHandlingService processFailureHandlingService;
 
-    public boolean handleProcessingError(Exception exception, TaskDefinition taskDefinition) {
+    public boolean handleProcessingError(Throwable exception, TaskDefinition taskDefinition) {
+
+        if (ExceptionUtils.getRootCause(exception) instanceof GpcServerErrorException) {
```

Comment:
Hmm, given this code is a bit nasty, what's the value in even having the `RetryLimitReachedException`. Why not throw a `GpcServerErrorException` instead?

Reply:
The GpcServerErrorException triggers the retry. When the retry fails it is given as the root cause. This is necessary because the MHS client can also throw a RetryLimitReachedException and we need to send the correct error code to the GP2GP Requester.